opentelemetry-api/opentelemetry/trace/__init__.py

```diff
@@ -231,3 +232,37 @@ class TraceOptions(int):
 # TODO
 class TraceState(typing.Dict[str, str]):
     pass
+
+
+_TRACER = None
+
+def tracer() -> Tracer:
+    """Gets the current global :class:`~.Tracer` object.
+    If there isn't one set yet, a default will be loaded."""
+
+    global _TRACER #pylint:disable=global-statement
+
+    if _TRACER is None:
+        #pylint:disable=protected-access
+        _TRACER = loader._load_impl(Tracer)
+
+    return _TRACER
+
+def set_preferred_tracer_implementation(
+        factory: typing.Callable[[typing.Type[Tracer]], Tracer]) -> None:
+    """Sets a callback which to query for the tracer implementation.
+
+    See :mod:`opentelemetry.loader` for details.
+
+    This function may not be called after a tracer is already loaded.
+
+    Args:
+            factory: A function that, when called with the :class:`Tracer` type
+                as an argument, returns an instance of :class:`Tracer`.
+    """
+    if _TRACER:
+        raise RuntimeError("Tracer already loaded.")
+
+    #pylint:disable=protected-access
+    loader._CALLBACKS_BY_TYPE[Tracer] = factory
```

Comment:
Setting the tracer factory in the trace class modifies the callbacks in `loader`? I'd expect it to do something like set a module-level constant instead and only fall back to the loader if that was never set. But that only works if we treat configuration in code as higher priority than env vars.

Since the tracer is set lazily it's possible to set this factory multiple times, the last call would win. Is that intentional?

Reply:
Actually a good point. Thinking about it, the dictionary in `loader.py` is uselessly complicated. The only advantage is that one can call `loader._load_impl` directly instead of using `trace.tracer()` but I guess that isn't really buying us much.

Reply:
I removed the dictionary.

Regarding calling the setter multiple times: I have no strong opinion on that. I thought maybe some vendor may call `set_preferred_impl` themselves and then the application would have the final say if it called that setter again. Besides, the current behavior is the simplest to implement. 😄